File: anagram.py
#!/usr/bin/env python2.7

import itertools
import re
import sys

ANAGRAM_LENGTH = 2
SET_LENGTH = None
MINIMUM_WORD_LENGTH = 4
KEY_LENGTH = 1

class Corpus(object):
  def __init__(self):
    self._words = []

  def stdin(self):
    for word in re.sub('[\W\d]+', ' ', sys.stdin.read()).split():
      self.add_word(word)

  def add_word(self, word):
    if len(word) < MINIMUM_WORD_LENGTH:
      return
    word = word.lower()
    for w in self._words:
      if w == word: return
    self._words.append(word)

  def find_matches(self, anagram_length=ANAGRAM_LENGTH, set_length=SET_LENGTH):
    matches = {}
    data = {}
    match_length = 2
    for pair in itertools.combinations(self._words, anagram_length):
      s = set(pair)
      t = ''.join(sorted(''.join(pair)))
      key = t[:KEY_LENGTH]
      if not key in data: data[key] = {}
      if not t in data[key]: data[key][t] = []
      if self._is_unique(s, data[key][t]):
        data[key][t].append(s)
        matches[t] = len(data[key][t])
        if matches[t] > match_length: match_length = matches[t]
    if not set_length:
      print "(length is {})".format(match_length)
    else:
      match_length = set_length
    for t in matches:
      if matches[t] != match_length: continue
      for n in data[t[:KEY_LENGTH]][t]:
        sys.stdout.write("{}; ".format(', '.join(n)))
      print

  def _is_unique(self, pair, matches):
    for possible in matches:
      if (possible & pair) != 0:
        return False
    return True

if __name__ == '__main__':
  corpus = Corpus()
  corpus.stdin()
  corpus.find_matches()

File: anagram.rb
#!/usr/bin/env ruby

ANAGRAM_LENGTH = 2
SET_LENGTH = nil
MINIMUM_WORD_LENGTH = 4
KEY_LENGTH = 1

class Corpus
  def initialize()
    @words = []
  end

  def stdin()
    $stdin.read.gsub(/[\W\d]+/, ' ').downcase.split.each do |word|
      add_word(word)
    end
  end

  def add_word(word)
    return if word.length < MINIMUM_WORD_LENGTH or @words.include?(word)
    @words << word
  end

  def find_matches(anagram_length=ANAGRAM_LENGTH, set_length=SET_LENGTH)
    matches = {}
    data = {}
    match_length = 2
    @words.combination(anagram_length) do |pair|
      composition = pair.join.each_char.to_a.sort.join
      key = composition[0, KEY_LENGTH]
      data[key] = {} unless data[key]
      data[key][composition] = [] unless data[key][composition]
      unique = true
      data[key][composition].each do |existing|
        if existing & pair != []
          unique = false
          break
        end
      end
      if unique
        data[key][composition] << pair
        matches[composition] = data[key][composition].length
        match_length = matches[composition] if matches[composition] > match_length
      end
    end
    unless set_length
      puts "(length is #{match_length})"
    else
      match_length = set_length
    end
    matches.each_pair do |composition,length|
      next if length != match_length
      data[composition[0, KEY_LENGTH]][composition].each do |pair|
        print "#{pair.join(', ')}; "
      end
      puts
    end
  end
end

if __FILE__ == $0
  corpus = Corpus.new
  corpus.stdin()
  corpus.find_matches()
end
